src/osd/ReplicatedBackend.cc

```diff
@@ -1667,35 +1688,116 @@ void ReplicatedBackend::submit_push_data(
   ObjectStore::Transaction *t)
 {
   hobject_t target_oid;
+  bool need_remove = false;
+  bool need_clone = false;
   if (first && complete) {
     target_oid = recovery_info.soid;
+    if (recovery_info.copy_subset.contains(0, recovery_info.size)) {
+      assert(recovery_info.copy_subset.size() == recovery_info.size);
+      need_remove = true;
+    }
   } else {
     target_oid = get_parent()->get_temp_recovery_object(recovery_info.version,
 							recovery_info.soid.snap);
     if (first) {
       dout(10) << __func__ << ": Adding oid "
 	       << target_oid << " in the temp collection" << dendl;
       add_temp_obj(target_oid);
+      need_remove = true;
+      need_clone = true;
     }
   }
 
   if (first) {
-    t->remove(coll, ghobject_t(target_oid));
-    t->touch(coll, ghobject_t(target_oid));
+    if (need_remove) {
+      t->remove(coll, ghobject_t(target_oid));
+      t->touch(coll, ghobject_t(target_oid));
+      bufferlist bv = attrs[OI_ATTR];
+      object_info_t oi(bv);
+      t->set_alloc_hint(coll, ghobject_t(target_oid),
+			oi.expected_object_size,
+			oi.expected_write_size,
+			oi.alloc_hint_flags);
+    } else {
+      //remove xattr and update later if overwrite on original object
+      t->rmattrs(coll, ghobject_t(target_oid));
+      //if need update omap, clear the previous content first
+      if (!before_progress.omap_complete)
+        t->omap_clear(coll, ghobject_t(target_oid));
+    }
     t->truncate(coll, ghobject_t(target_oid), recovery_info.size);
     t->omap_setheader(coll, ghobject_t(target_oid), omap_header);
 
-    bufferlist bv = attrs[OI_ATTR];
-    object_info_t oi(bv);
-    t->set_alloc_hint(coll, ghobject_t(target_oid),
-		      oi.expected_object_size,
-		      oi.expected_write_size,
-		      oi.alloc_hint_flags);
+    if (need_clone) {
+      // need preserver local omap content if omap is undirty
+      int r;
+      if (before_progress.omap_complete) {
+        map<string, bufferlist> out;
+        r = store->omap_get(coll, ghobject_t(recovery_info.soid), NULL, &out);
+        if (r == 0) {
+          t->omap_setkeys(coll, ghobject_t(target_oid), out);
+        } else if (r == -ENOENT) {
+          dout(10) << " local object " << recovery_info.soid << " does not exist "
+             <<" skip retrieve object map" << dendl;
+        } else {
+          derr << "tried to get object map from local object " << recovery_info.soid
+               << " but got " << cpp_strerror(-r) << "\n" << dendl;
+          assert(0 == "fail to retrieve object map");
+        }
+      }
+      //clone overlap content in local object
+      struct stat st;
+      r = store->stat(coll, ghobject_t(recovery_info.soid), &st);
```

Comment:
Yes, you are going to have to maintain information about whether the target osd has the object -- presumably in the missing item just as you are maintaining the clean_regions etc.